File: frontend/src/api/collect.ts
/**
 * @Author: li
 * @Email: lijianqiao2906@live.com
 * @FileName: collect.ts
 * @DateTime: 2026-01-10
 * @Docs: ARP/MAC 采集 API 模块
 */

import { request } from '@/utils/request'
import type { ResponseBase } from '@/types/api'

// ==================== 接口定义 ====================

/** ARP 表条目 */
export interface ARPEntry {
  ip_address: string
  mac_address: string
  interface?: string | null
  vlan_id?: string | null
  entry_type?: string | null
  updated_at?: string | null

  // 兼容旧字段
  vlan?: string | null
  type?: string | null
}

/** MAC 表条目 */
export interface MACEntry {
  mac_address: string
  vlan_id?: string | null
  interface?: string | null
  entry_type?: string | null
  state?: string | null
  updated_at?: string | null

  // 兼容旧字段
  vlan?: string | null
  type?: string | null
}

/** ARP 表响应 */
export interface ARPTableResponse {
  device_id: string
  device_name: string | null
  entries: ARPEntry[]
  total: number
  cached_at: string | null
}

/** MAC 表响应 */
export interface MACTableResponse {
  device_id: string
  device_name: string | null
  entries: MACEntry[]
  total: number
  cached_at: string | null
}

/** 设备采集结果 */
export interface DeviceCollectResult {
  device_id: string
  device_name: string | null
  success: boolean
  arp_count: number
  mac_count: number
  error_message: string | null
  duration_ms: number | null
}

/** 批量采集结果 */
export interface CollectResult {
  total_devices: number
  success_count: number
  failed_count: number
  results: DeviceCollectResult[]
  started_at: string | null
  completed_at: string | null
}

/** 采集任务状态 */
export interface CollectTaskStatus {
  task_id: string
  status: 'PENDING' | 'STARTED' | 'SUCCESS' | 'FAILURE'
  progress: number | null
  result: CollectResult | null
  error: string | null
}

/** 定位响应 */
export interface LocateResponse {
  ip_address: string | null
  mac_address: string
  device_id: string | null
  device_name: string | null
  interface: string | null
  vlan: string | null
  located_at: string | null
}

/** 批量采集请求 */
export interface BatchCollectRequest {
  device_ids: string[]
  collect_arp?: boolean
  collect_mac?: boolean
  otp_code?: string
}

/** 单设备采集请求 */
export interface CollectDeviceRequest {
  collect_arp?: boolean
  collect_mac?: boolean
  otp_code?: string
}

// ==================== API 函数 ====================

/** 手动采集单设备 */
export function collectDevice(deviceId: string, data: CollectDeviceRequest = {}) {
  return request<ResponseBase<DeviceCollectResult>>({
    url: `/collect/device/${deviceId}`,
    method: 'post',
    data,
  })
}

/** 批量采集设备 */
export function batchCollect(data: BatchCollectRequest) {
  return request<ResponseBase<CollectResult>>({
    url: '/collect/batch',
    method: 'post',
    data,
  })
}

/** 异步批量采集 */
export function batchCollectAsync(data: BatchCollectRequest) {
  return request<ResponseBase<CollectTaskStatus>>({
    url: '/collect/batch/async',
    method: 'post',
    data,
  })
}

/** 查询采集任务状态 */
export function getCollectTaskStatus(taskId: string) {
  return request<ResponseBase<CollectTaskStatus>>({
    url: `/collect/task/${taskId}`,
    method: 'get',
  })
}

/** 获取设备 ARP 表 */
export function getDeviceARPTable(deviceId: string) {
  return request<ResponseBase<ARPTableResponse>>({
    url: `/collect/device/${deviceId}/arp`,
    method: 'get',
  })
}

/** 获取设备 MAC 表 */
export function getDeviceMACTable(deviceId: string) {
  return request<ResponseBase<MACTableResponse>>({
    url: `/collect/device/${deviceId}/mac`,
    method: 'get',
  })
}

/** IP 地址定位 */
export function locateIP(ipAddress: string) {
  return request<ResponseBase<LocateResponse>>({
    url: `/collect/locate/ip/${encodeURIComponent(ipAddress)}`,
    method: 'get',
  })
}

/** MAC 地址定位 */
export function locateMAC(macAddress: string) {
  return request<ResponseBase<LocateResponse>>({
    url: `/collect/locate/mac/${encodeURIComponent(macAddress)}`,
    method: 'get',
  })
}

File: backend/app/schemas/collect.py
"""
@Author: li
@Email: lijianqiao2906@live.com
@FileName: collect.py
@DateTime: 2026-01-09 22:00:00
@Docs: ARP/MAC 采集 Pydantic Schema 定义。
"""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from app.utils.validators import validate_ip_address, validate_mac_address

# ===== ARP 表相关 =====


class ARPEntry(BaseModel):
    """ARP 表条目。"""

    ip_address: str = Field(..., description="IP 地址")
    mac_address: str = Field(..., description="MAC 地址")
    vlan_id: str | None = Field(default=None, description="VLAN ID")
    interface: str | None = Field(default=None, description="接口名称")
    age: str | None = Field(default=None, description="老化时间")
    entry_type: str | None = Field(default=None, description="条目类型 (Dynamic/Static)")
    updated_at: datetime = Field(default_factory=datetime.now, description="更新时间")

    @field_validator("ip_address")
    @classmethod
    def validate_ip(cls, v: str) -> str:
        """验证 IP 地址格式。"""
        return validate_ip_address(v)

    @field_validator("mac_address")
    @classmethod
    def validate_mac(cls, v: str) -> str:
        """验证 MAC 地址格式。"""
        return validate_mac_address(v)


class ARPTableResponse(BaseModel):
    """ARP 表响应。"""

    device_id: UUID = Field(..., description="设备ID")
    device_name: str | None = Field(default=None, description="设备名称")
    entries: list[ARPEntry] = Field(default_factory=list, description="ARP 条目列表")
    total: int = Field(default=0, description="总条目数")
    cached_at: datetime | None = Field(default=None, description="缓存时间")


# ===== MAC 表相关 =====


class MACEntry(BaseModel):
    """MAC 地址表条目。"""

    mac_address: str = Field(..., description="MAC 地址")
    vlan_id: str | None = Field(default=None, description="VLAN ID")
    interface: str | None = Field(default=None, description="接口名称")
    entry_type: str | None = Field(default=None, description="条目类型 (Learned/Config/Static)")
    state: str | None = Field(default=None, description="状态")
    updated_at: datetime = Field(default_factory=datetime.now, description="更新时间")

    @field_validator("mac_address")
    @classmethod
    def validate_mac(cls, v: str) -> str:
        """验证 MAC 地址格式。"""
        return validate_mac_address(v)


class MACTableResponse(BaseModel):
    """MAC 地址表响应。"""

    device_id: UUID = Field(..., description="设备ID")
    device_name: str | None = Field(default=None, description="设备名称")
    entries: list[MACEntry] = Field(default_factory=list, description="MAC 条目列表")
    total: int = Field(default=0, description="总条目数")
    cached_at: datetime | None = Field(default=None, description="缓存时间")


# ===== 采集请求 =====


class CollectDeviceRequest(BaseModel):
    """采集单设备请求。"""

    collect_arp: bool = Field(default=True, description="是否采集 ARP 表")
    collect_mac: bool = Field(default=True, description="是否采集 MAC 表")
    otp_code: str | None = Field(default=None, description="OTP 验证码（如果设备需要）")


class CollectBatchRequest(BaseModel):
    """批量采集请求。"""

    device_ids: list[UUID] = Field(..., min_length=1, description="设备ID列表")
    collect_arp: bool = Field(default=True, description="是否采集 ARP 表")
    collect_mac: bool = Field(default=True, description="是否采集 MAC 表")
    otp_code: str | None = Field(default=None, description="OTP 验证码（如果设备需要）")


# ===== 采集结果 =====


class DeviceCollectResult(BaseModel):
    """单设备采集结果。"""

    device_id: UUID = Field(..., description="设备ID")
    device_name: str | None = Field(default=None, description="设备名称")
    success: bool = Field(default=False, description="是否成功")
    arp_count: int = Field(default=0, description="ARP 条目数")
    mac_count: int = Field(default=0, description="MAC 条目数")
    error_message: str | None = Field(default=None, description="错误信息")
    duration_ms: int | None = Field(default=None, description="耗时（毫秒）")


class CollectResult(BaseModel):
    """采集任务结果。"""

    task_id: str | None = Field(default=None, description="Celery 任务ID（异步任务时返回）")
    total_devices: int = Field(default=0, description="总设备数")
    success_count: int = Field(default=0, description="成功数")
    failed_count: int = Field(default=0, description="失败数")
    results: list[DeviceCollectResult] = Field(default_factory=list, description="各设备采集结果")
    started_at: datetime | None = Field(default=None, description="开始时间")
    completed_at: datetime | None = Field(default=None, description="完成时间")


class CollectTaskStatus(BaseModel):
    """采集任务状态（用于查询异步任务）。"""

    task_id: str = Field(..., description="任务ID")
    status: str = Field(..., description="任务状态 (PENDING/STARTED/SUCCESS/FAILURE)")
    progress: int = Field(default=0, description="进度百分比")
    result: CollectResult | None = Field(default=None, description="任务结果（完成时）")
    error: str | None = Field(default=None, description="错误信息（失败时）")


# ===== IP/MAC 定位相关 =====


class LocateMatch(BaseModel):
    """单条定位匹配结果。"""

    device_id: UUID = Field(..., description="设备ID")
    device_name: str | None = Field(default=None, description="设备名称")
    device_ip: str | None = Field(default=None, description="设备管理IP")
    interface: str | None = Field(default=None, description="接口/端口")
    vlan_id: str | None = Field(default=None, description="VLAN ID")
    ip_address: str | None = Field(default=None, description="IP 地址（MAC 查询时返回）")
    mac_address: str | None = Field(default=None, description="MAC 地址（IP 查询时返回）")
    entry_type: str | None = Field(default=None, description="条目类型")
    cached_at: datetime | None = Field(default=None, description="数据缓存时间")


class LocateResponse(BaseModel):
    """定位查询响应。"""

    query: str = Field(..., description="查询的 IP 或 MAC 地址")
    query_type: str = Field(..., description="查询类型 (ip/mac)")
    matches: list[LocateMatch] = Field(default_factory=list, description="匹配结果列表")
    total: int = Field(default=0, description="匹配总数")
    searched_devices: int = Field(default=0, description="搜索的设备数")
    search_time_ms: int = Field(default=0, description="搜索耗时（毫秒）")
